### PR TITLE
graphql: reworking Socket Client API to support custom header

### PR DESCRIPTION
**Pull request by vincenzopalazzo**
Fixes https://github.com/zino-hofmann/graphql-flutter/issues/815

- [X] Cleaning we test
- [ ] Adding custom header test, this required to change a little bit the mock server

### CODE REVIEW

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart [hunk 1]
```diff
@@ -86,30 +87,43 @@ class SocketClientConfig {
...
-  /// connect: (url, protocols) {
+  /// customConnect: (url, protocols) {
   ///    var channel = WebSocketChannel.connect(url, protocols: protocols)
   ///    // without this line, our client won't be able to listen to stream events,
   ///    // because you are already listening.
   ///    channel = channel.forGraphQL();
   ///    channel.stream.listen(myListener)
   ///    return channel;
-  /// }
+  /// }socketChannel
```
Comment (budde377) [inline packages/graphql/lib/src/links/websocket_link/websocket_client.dart:97]: Typo?

```suggestion
  /// }
```

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart [hunk 2]
```diff
@@ -86,30 +87,43 @@ class SocketClientConfig {
...
-  final WebSocketConnect connect;
-
-  static Future<WebSocketChannel> defaultConnect(
-    Uri uri,
-    Iterable<String>? protocols,
-  ) async {
+  final WebSocketConnect? customConnect;
+
+  /// Custom header to add inside the client
+  final Map<String, dynamic>? customHeaders;
```
Comment (budde377) [inline packages/graphql/lib/src/links/websocket_link/websocket_client.dart:108]: When is this used? And maybe:

```suggestion
  final Map<String, dynamic>? headers;
```

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart [hunk 3]
```diff
@@ -86,30 +87,43 @@ class SocketClientConfig {
...
+
+  /// Function to define another connection without call directly
+  /// the connection function
+  FutureOr<WebSocketChannel> connect(
+      {required Uri uri,
+      Iterable<String>? protocols,
+      Map<String, dynamic>? headers,
+      WebSocketConnect? fnConnect}) {
```
Comment (budde377) [inline packages/graphql/lib/src/links/websocket_link/websocket_client.dart:116]:
```suggestion
      WebSocketConnect? connectFn}) {
```

WDYT?

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart [hunk 4]
```diff
@@ -86,30 +87,43 @@ class SocketClientConfig {
   /// Warning: if you want to listen to the listen to the stream,
   /// wrap your channel with our [GraphQLWebSocketChannel] using the `.forGraphQL()` helper:
   /// ```dart
-  /// connect: (url, protocols) {
+  /// customConnect: (url, protocols) {
```
Comment (budde377) [inline packages/graphql/lib/src/links/websocket_link/websocket_client.dart:90]: Do you think we should be consistent with the naming of this and the `fnConnect` function? That way, people will know that the functions basically do the same.

**Review by budde377 — COMMENTED** [review]
Does it make sense to pass the headers to the connect function? 

```dart
typedef WebSocketConnect = FutureOr<WebSocketChannel> Function(
  Uri uri,
  Map<String, dynamic> headers,
  Iterable<String>? protocols,
);
```

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart (hunk 2 above)
Comment (vincenzopalazzo) [reply packages/graphql/lib/src/links/websocket_link/websocket_client.dart:108]: It is used only we are not using flutter when, in particular in this function

```
   return defaultConnectPlatform(
      uri,
      protocols,
      headers: headers,
    );
```

P.S: I Like the `header` name, thanks

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart (hunk 4 above)
Comment (vincenzopalazzo) [reply packages/graphql/lib/src/links/websocket_link/websocket_client.dart:90]: I agree, I will move all under the name `connectFn`

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart [hunk 5]
```diff
@@ -40,16 +40,17 @@ class SubscriptionListener {
 
 enum SocketConnectionState { notConnected, connecting, connected }
 
-class SocketClientConfig {
-  const SocketClientConfig({
+class customConnect {
```
Comment (budde377) [inline packages/graphql/lib/src/links/websocket_link/websocket_client.dart:43]: Typo?

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart [hunk 6]
```diff
@@ -95,21 +96,28 @@ class SocketClientConfig {
...
+    if (connectFn != null) {
+      return connectFn(uri, protocols);
+    }
+    if (this.connectFn != null) {
+      return this.connectFn!(uri, protocols);
+    }
     return defaultConnectPlatform(
       uri,
       protocols,
+      headers: headers,
```
Comment (budde377) [inline packages/graphql/lib/src/links/websocket_link/websocket_client.dart:120]: If I'm reading this correctly, the `headers` property is never used. Shouldn't this be something like

```suggestion
      headers: headers ?? this.headers,
```

**Comment by budde377** [comment]
> The only reason to support the custom function in the connect function is that can exist a use case where it is good to create a unique connection without recreating a new Websocket client.

We can keep it as is if you prefer, if nothing else it'll preserve backward compatibility.

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart (hunk 5 above)
Comment (vincenzopalazzo) [reply packages/graphql/lib/src/links/websocket_link/websocket_client.dart:43]: Ops! Sorry about that and thanks for the catch!

File: packages/graphql/lib/src/links/websocket_link/websocket_client.dart (hunk 6 above)
Comment (vincenzopalazzo) [reply packages/graphql/lib/src/links/websocket_link/websocket_client.dart:120]: Yes, Good catch thanks

**Comment by vincenzopalazzo** [comment]
>We can keep it as is if you prefer, if nothing else it'll preserve backward compatibility.

I reflected on this a little bit, and from the actual API, I think it is more users have the function in the SocketClientConf. let's go in this way and after we will see if this require more work

**Review by budde377 — APPROVED** [review]

**Pull request merged**